_posts/2025-11-06-sealing.markdown

```diff
@@ -0,0 +1,202 @@
+
+Our virtualised mechanism takes advantage of the fact that we can represent a lot more values in the address field of a capability than in the otype field.
+Any address in a permit-seal or permit-unseal capability that cannot fit into the otype field is available to authorise sealing or unsealing with the virtualised mechanism.
+This gives us almost four billion types that we can represent with the virtualised mechanism, at the expense of being able to seal only complete objects.
+This is fine for most use cases, though language VMs on bigger systems would benefit from at least a few hardware otypes.
+
+The CHERIoT `token_unseal` function unseals the object using the hardware sealing key and then compares the virtual otype in the header to the type of the permit-unseal capability passed as the key.
+If they match, it returns an unsealed capability to the object without the header, so the caller doesn't ever have access to the header.
+
+Sealed pointers using the software sealing mechanism always point to the end of the header.
+The header must be strongly aligned, which means that the low three bits are unused.
+We use these to implement three *software permissions*.
+As with the CHERI permissions, these can be cleared but not set.
+
+# Type safety gives easy-to-use handles
+
+A lot of CHERIoT compartments use type-safe pointers as handles.
+For example, if you open a socket, you get back a sealed capability to that socket's state.
+The same thing happens for message queues, TLS sessions, and so on.
+
+In a conventional monolithic kernel, these would all be handles or file descriptors looked up in some table in the kernel.
+In a microkernel, they'd either be indexed per caller, or managed via some handle-manager service.
+In CHERIoT, these are simply opaque pointers.
+
+Exactly the same abstractions that you use for data hiding in good API design work between trust domains, with a little bit of hardware help.
+This is great for programmers because you can easily *retrofit* a security boundary.
+If you have built an API around exposing opaque types, turning it into a robust security boundary simply means doing a few checks on the public APIs.
+
+# Type safety gives static software-defined capabilities
+
+The same sealing mechanism that provides type safety for dynamic allocations also works for static objects.
+You can create a CHERIoT compartment that has access to a sealed object where another compartment owns the sealing key.
+The contents of these show up in the audit log.
+
+We use these throughout the system to implement *software* capabilities.
+CHERI (hardware) capabilities are unforgeable (delegable) tokens of authority that authorise access to memory.
+CHERIoT software capabilities are unforgeable (delegable) tokens of authority that authorise some other compartment to perform some action on your behalf.
```

Comment:
```suggestion
CHERI (hardware) capabilities are unforgeable (delegable) tokens of authority to perform some _architectural_ action; they authorize operations carried out on your behalf by the hardware.
CHERIoT software capabilities are unforgeable (delegable) tokens of authority to perform some _software_ action; they authorise operations carried out on your behalf by some other compartment.
```

_In principle_ our encoding admits sealed sealing-root-derived caps.  We've never used those and I don't imagine why we would, but...